### DISCUSSION TITLE
Autodevices at lower bit depths (Tesla P40 vs 30-series, FP16, int8, and int4)

### DISCUSSION THREAD

**Discussion by tensiondriven** [oobabooga/text-generation-webui#1701] [Q&A]
Hola - I have a few questions about older Nvidia Tesla cards.  These questions have come up on Reddit and elsewhere, but there are a couple of details that I can't seem to get a firm answer to.

I guess the main question is: Does the Tesla P40's lack of floating-point hamper performance for int8 or int4 models because of it's lack of floating point?

Said differently, just how terrible of an idea is it to load up a machine with P40's, assuming you're ONLY doing int8 or int4 INFERENCE?

For context, I'm currently running dual 3090's on a motherboard that has one PCIe slot limited to Gen 3 x 4.  I typically run llama-30b in 4bit, no groupsize, and it fits on one card.  It's probably generating 6-8 tokens/sec if I had to guess.  When I use autodevices and spill over to the second card, performance drops down to 2-3 tokens/sec.  I think this is partially (mostly? all?) because of the Gen3x4 pci bus that the second card is using.

I realize that 30-series cards have 2x the memory bandwidth and can utilize much faster pci bus. I also know that the P40 has limited floating point functionality, and the 30-series has the addition of Tensor Cores, but it isn't clear to me how much those Tensor Cores improve inference at 8bit and 4bit.

What also isn't clear to me is when these low-bit-depth models need to be blown up to higher bit depths, and how that affeects performance in multigpu configurations.  In the case of int4 and int8, does data travel over the pcie bus at 4bit/8bit respectively, or is it embiggened when it leaves the GPU?

If it's true that my $2000 worth of 3090's is rate-limited by the PCI bus, would I do just as well to replace the 3090's with P40's in this specific machine, assuming an autodevices workload and 8bit models?  What about 4bit, are they treated differently?

Does the conversion to/from int4/int8 require special cores that are present on the 30-series but not the P40?  

And, is int4 a special beast, distinct from int8?  That is to ask, do int4 models have vastly different performance characteristics compared to int8, or is it truly a 2-for-1 kinda deal, inside the GPU, on the pci bus, in system memory, etc?

I hope these questions makes sense, I can't find anything online that addresses these nuances.  Thanks in advanced!

PS Bigup to the amazing [Tim Dettmers](https://timdettmers.com/) for his work on bitsandbytes, and his great blog that goes way back to 2014.  This article was especially interesting:

https://timdettmers.com/2022/08/17/llm-int8-and-emergent-features/

**Comment by cduk** [comment]
For inference, the pcie bandwidth is not a limit. I just bought a P40 and tested it. I'm currently running with power limit of 125W (half rated) as I have no proper cooling for it yet. It is actually thermally throttling to around 80W-90W. Using a Qwen 7B model with Int8 quantization, I'm getting round 25 tokens per second - which is pretty decent esp. considering the power/thermal throttling.
